daft/sql/sql_connection.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+import logging
+import warnings
+from typing import TYPE_CHECKING, Callable
+from urllib.parse import urlparse
+
+import pyarrow as pa
+
+if TYPE_CHECKING:
+    from sqlalchemy.engine import Connection
+
+logger = logging.getLogger(__name__)
+
+
+class SQLConnection:
+    def __init__(self, conn: str | Callable[[], Connection], driver: str, dialect: str) -> None:
+        self.conn = conn
+        self.dialect = dialect
+        self.driver = driver
+
+    def __repr__(self) -> str:
+        return f"SQLConnection(conn={self.conn})"
+
+    @classmethod
+    def from_url(cls, url: str) -> SQLConnection:
+        scheme = urlparse(url).scheme.strip().lower()
+        if "+" in scheme:
+            dialect, driver = scheme.split("+")
+        else:
+            dialect, driver = scheme, ""
+        return SQLConnection(url, driver, dialect)
+
+    @classmethod
+    def from_connection_factory(cls, conn_factory: Callable[[], Connection]) -> SQLConnection:
+        try:
+            with conn_factory() as connection:
+                if not hasattr(connection, "engine"):
```

Comment:
Should this be a `isinstance(connection, SQLAlchemy.Connection)` check instead of a hasattr?

Reply:
Yeah that should work!